### DISCUSSION TITLE
Requested user is not a administrator. Allow the user to administer this computer in System Preferences to continue.

### DISCUSSION THREAD

**Discussion by Wooster101** [grahampugh/erase-install#81] [Q&A]
Hello and thanks for your excellent work with this script!

I have troubles making this script work on M1 Macs

I run the script from jamf in a policy from Self Service. The full-screen jamfHelper comes up for a minute and then goes away.

In the beginning of the logging I get that the user is an admin and owner of the volume - later on I get the user is not an admin (it still is)

Log below:

```
Script result: 
[erase-install] v0.19.0 script execution started: Mon Apr 12 15:08:09 CEST 2021
[erase-install] Caffeinating this script (pid=17285)
45000000
[free_space_check] OK - 180356512 KB free disk space detected
[erase-install] Looking for existing installer
[find_existing_installer] Installer sparse image found at /Library/Management/erase-install/Install_macOS_11.2.3-20D91.sparseimage.
/dev/disk4          	GUID_partition_scheme          	
/dev/disk4s1        	EFI                            	
/dev/disk4s2        	Apple_HFS                      	/Volumes/Install_macOS_11.2.3-20D91
   [check_installer_is_valid] Checking validity of /Volumes/Install_macOS_11.2.3-20D91/Applications/Install macOS Big Sur.app.
Volume Shared Support on disk5s2 force-unmounted
[check_installer_is_valid] Using Build value from com_apple_MobileAsset_MacSoftwareUpdate.xml
Volume Shared Support on disk5s2 force-unmounted
[check_installer_is_valid] Installer: 20D91 ; System: 20D91 : valid build.
yes user.test is a member of staff 
yes user.test is a member of admin 
[get_user_details] user.test is a Volume Owner 
[check_password] Success: the password entered is the correct login password for user.test. 
[erase-install] WARNING! Running /Volumes/Install_macOS_11.2.3-20D91/Applications/Install macOS Big Sur.app with eraseinstall option 
[erase-install] Opening jamfHelper full screen message (language=en) 
[get_user_details] updating preboot files (takes a few seconds)... 
By using the agreetolicense option, you are agreeing that you have run this tool with the license only option and have read and agreed to the terms. If you do not agree, press CTRL-C and cancel this process immediately. 
Requested user is not a administrator. Allow the user to administer this computer in System Preferences to continue.
[erase-install] 'Self Service' ended
[erase-install] 'jamfHelper' ended
/Library/Application Support/JAMF/tmp/ZZZZ TEST ZZZZ - Erase-Install macOS: line 148: 17487 Terminated: 15          "$jamfHelper" -windowType $window_type -title "${!dialog_erase_title}" -heading "${!dialog_erase_title}" -description "${!dialog_erase_desc}" -icon "$dialog_erase_icon"
 [erase-install] 'caffeinate' ended
```

have you got any ideas what I can try next?

Best regards
Holger

**Comment by grahampugh** [comment]
@Wooster101 How did you create `user.test`?

Please can you try the following:

1. `ls /System/Volumes/Preboot`
2. There should be a folder with a name like a UUID e.g. `ACF62396-59F0-4234-9038-33493694F231` in my case. 
3. `defaults read /System/Volumes/Preboot/[name of folder]/var/db/AdminUserRecoveryInfo.plist`

Is `user.test` listed in this file?  If not, what is the entry for this user in `defaults read /System/Volumes/Preboot/[name of folder]/var/db/AllUsersInfo.plist`?

**Reply by Wooster101** [reply]
Thanks for your very fast reply!

- The user.test is created as the first user in Setup Assistant.

- This user has SecureToken when I check. FileVault is enabled and both this user.test and a hidden admin has Secure Token and are able to unlock FileVault.

- I have found the file but running default read as user,  sudo or root on this result in:
Domain /System/Volumes/Preboot/[name of folder]/var/db/AdminUserRecoveryInfo.plist does not exist (correct folder given in path ;) )

- Running cat on the path above shows a empty Apple plist file

- running defaults read on AllUsersInfo.plist shows the user.test being admin, AuthenticationAuthority (ShadowHash, Kerberos V5, securetoken), GUID, InputSources, RealName and UserPictureData information.

**Reply by grahampugh** [reply]
Thanks. And does the output of those files change if you run the following command? :-

```
diskutil apfs updatePreboot /
```

**Reply by grahampugh** [reply]
This is the kind of output I would expect in `AdminUserRecoveryInfo.plist`:

```
    test =     {
        AuthenticationAuthority =         (
            ";ShadowHash;HASHLIST:<SALTED-SHA512-PBKDF2,SRP-RFC5054-4096-SHA512-PBKDF2>",
            ";SecureToken;",
            ";Kerberosv5;;test@LKDC:SHA1.E9122C96BAB43FA6B.....5FBCB8F7AE87;LKDC:SHA1.E9122C96BAB43FA6BA0A322D200A5FBCB8F7AE87;"
        );
        GeneratedUID = "3EC4D455-DCEF-4BA2-B2CA-XXXXXXXXXX";
        RealName = test;
    };
```

And this in `AllUsersInfo.plist `:

```
    test =     {
        Admin = 1;
        AuthenticationAuthority =         (
            ";ShadowHash;HASHLIST:<SALTED-SHA512-PBKDF2,SRP-RFC5054-4096-SHA512-PBKDF2>",
            ";SecureToken;",
            ";Kerberosv5;;test@LKDC:SHA1.E9122C96BAB43FA6B.....5FBCB8F7AE87;LKDC:SHA1.E9122C96BAB43FA6BA0A322D200A5FBCB8F7AE87;"
        );
        GeneratedUID = "3EC4D455-DCEF-4BA2-B2CA-XXXXXXXXXX";
        InputSources = "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>\\n<!DOCTYPE plist PUBLIC \\"-//Apple//DTD PLIST 1.0//EN\\" \\"http://www.apple.com/DTDs/PropertyList-1.0.dtd\\">\\n<plist version=\\"1.0\\">\\n<array>\\n\\t<dict>\\n\\t\\t<key>InputSourceKind</key>\\n\\t\\t<string>Keyboard Layout</string>\\n\\t\\t<key>KeyboardLayout ID</key>\\n\\t\\t<integer>2</integer>\\n\\t\\t<key>KeyboardLayout Name</key>\\n\\t\\t<string>British</string>\\n\\t</dict>\\n</array>\\n</plist>\\n";
        RealName = test;
        UserPictureData = {length = 32813, bytes = 0xffd8ffe0 00104a46 49460001 01000048 ... 7a59b77f 5b9fffd9 };
    };
```

**Reply by Wooster101** [reply]
Hi again, 

Running this "diskutil apfs updatePreboot /" gives the following result:

AllUsersInfo.plist still looks the same

AdminUserRecoveryInfo.plist is still a empty Apple plist

I have rebooted and redone the update preboot couple of times with the same result.

**Reply by Wooster101** [reply]
My AllUsersInfo.plist looks the same as the posted above (with my unique info in it) except that my two users have an entry of Admin = 0; (as I would expect to mean admin as both my users are admins)

**Reply by grahampugh** [reply]
Interesting. Thanks for the tests. It does look like there needs to be an entry in `AdminUserRecoveryInfo.plist` for the restart to be authorised. So, it looks like there is yet another (as yet unknown) condition for allowing a user with a Secure Token to be a Volume Owner. Shame, we thought we had solved that one with v0.19.0.

Is the restart working using your hidden admin as the user? (I guess not if there's no `AdminUserRecoveryInfo.plist` entries).

**Reply by Wooster101** [reply]
Hidden admin does not work the times I have tried

- I now tried to create a new admin-user named aa in System Preferences

- This new user (aa) gets a 1 in AllUsersInfo.plist (Admin = 1;)

- Now the AdminUserRecoveryInfo.plist got data - BUT only this newly created admin-user aa is present. 

- This correlates with my findings before where I created a new admin-user (aa) and that user could run erase-install. 

- It seems like the old admin-users (that can do everything else) is not admin on the Preboot Volume as their data is Admin = 0;

**Reply by grahampugh** [reply]
OK, that's progress. We need an entry in the `AdminUserRecoveryInfo.plist` file to verify that the user is a volume owner. the question remains, how to make that happen!

**Reply by Wooster101** [reply]
The strange thing is that the test in the script gives a thumbs up for this when I read the log-file

[get_user_details] user.test is a Volume Owner

**Reply by grahampugh** [reply]
Not really strange, that's just my code interpreting the information I had that indicated that a user was a Volume Owner. It appears there are more criteria than the code checks for. 

None of this is documented by Apple.

**Reply by grahampugh** [reply]
@Wooster101 was your computer setup using DEP, or manually? And if you run `/usr/sbin/diskutil apfs listUsers /` manually does it show the user that isn't working as a Volume Owner?

**Reply by Wooster101** [reply]
The computer was setup using DEP

I will get back with the status in a couple of hours when back at my work where this test-computer is at the moment.

**Reply by Wooster101** [reply]
Here are the results of diskutil apfs listUsers /

All users are Volumes Owners (personal recovery, aa, user.test, hiddenadmin, bootstrap token)

Cryptographic users for disk3s5s1 (5 found)
|
+-- D5017E28-79F2-4920-BBEA-9C14632FAD37
|   Type: Local Open Directory User
|   Volume Owner: Yes
|
+-- 2457711A-523C-4604-B75A-F48A571D5036
|   Type: MDM Bootstrap Token External Key
|   Volume Owner: Yes
|
+-- EBC6C064-0000-11AA-AA11-00306543ECAC
|   Type: Personal Recovery User
|   Volume Owner: Yes
|
+-- 7C7848E0-C725-4FAB-8384-77AE200199DA
|   Type: Local Open Directory User
|   Volume Owner: Yes
|
+-- 7C8409B6-91C3-4026-89E9-FC3665D9B0FF
    Type: Local Open Directory User
    Volume Owner: Yes

**Reply by grahampugh** [reply]
I had some interaction with someone from Apple about this. They said it will be worth opening a Feedback ticket if this is reproducible. Any user listed as a Volume Owner via `diskutil apfs listUsers /` should be able to run `startosinstall --eraseinstall` according to them, and if not for some reason, they would expect `updatePreboot` to fix it, so let me know if you can reproduce this on another device or after a rebuild, and please open a feedback if so. I will relay the FR number to that person.

**Reply by Wooster101** [reply]
Hi, 

I'm in the process of erase and reinstalling this computer now - I used the script for this thru Self Service with the user aa (created above) logged in (this user was created in System Preferences that is an admin and is present in AdminUserRecoveryInfo.plist after the creation.)

I will try again after re-enrolling the computer with the user created at Setup Assistant (DEP enroll)

I'm sure it will behave the same as I have done this a couple of times already.

**Reply by Wooster101** [reply]
I have now re-enrolled the computer and the status is the same as before

After DEP enroll the admin created at Setup Assistant and the hidden admin (created by Jamf) are admins I every aspect except for not being flagged as admins in AllUsersInfo.plist and the file AdminUserRecoveryInfo.plist is an empty Apple plist

Creating a local admin (aa) thru System Preferences will after running updatePreboot populate the file AdminUserRecoveryInfo.plist and is flagged as an admin in AllUsersInfo.plist

I also created a local admin (bb) thru sysadminctl -addUser and that user is also listed as admin everywhere and is a Volume Owner (same as the user (aa) created in GUI)

I have created all the files Apple are asking for in their feedback system.

**Reply by Wooster101** [reply]
Here is the Apple FB number: FB9078997

**Reply by grahampugh** [reply]
Thanks, I've passed on the FB to the contact at Apple.

**Reply by Wooster101** [reply]
Hello,

Have you heard anything about this? I have had no response from Apple to my ticket yet.

**Reply by grahampugh** [reply]
Unfortunately not.

**Comment by Wooster101** [comment]
Hello, 

I see that the script is updated. I have not tried this new version. I suspect that it will not affect the missing admins in "AdminUserRecoveryInfo.plist"

I have not heard back from Apple regarding this. I have "poked" them again today.

**Reply by Wooster101** [reply]
Hello again, 

I have tried the last version of the script now (0.20.1) and the Erase and Install have worked 3 times on the same computer that was failing before. 

The Mac I test with is a MBA M1 and now have macOS 11.3.1 installed. It is the same computer I have tried with all the time but it had macOS 11.2.3 the last time I tried with version 0.19.0.

Just wanted to get back with my results :)

**Reply by grahampugh** [reply]
Good to hear! Thanks for the report.

**Comment by MikePattie** [comment]
Hey guys,

Im having this issue using the latest version of the script (26.1). Trying to update from macOS 11.6.7 to macOS 12.4.....Any updated ideas or fixes? I saw that it worked for Wooster101 on script version 0.20.1 and just wondering if this has been an ongoing issue throughout the versions?

Thanks!

**Reply by grahampugh** [reply]
Please take a look at https://github.com/grahampugh/erase-install/wiki/FAQs#the-installer-downloads-but-then-something-fails-why
